### ISSUE TITLE
broken request.values in Flask 2 on GET?

### ISSUE THREAD

**Issue by zx80** [pallets/flask#4120]
It seems that `request.values` is broken:

```python
from flask import Flask, request
app = Flask(__name__)

@app.get("/test")
def get_test():
        res = f"args: {request.args}\nform: {request.form}\nvalues: {request.values}\n"
        return res, 200
```

Then:

```shell
curl -X GET -i -d i=3 -d s=hello  http://0.0.0.0:5000/test
```

returns:
```http
HTTP/1.0 200 OK
Content-Type: text/html; charset=utf-8
Content-Length: 136
Server: Werkzeug/2.0.1 Python/3.8.5
Date: Tue, 01 Jun 2021 06:52:28 GMT

args: ImmutableMultiDict([])
form: ImmutableMultiDict([('i', '3'), ('s', 'hello')])
values: CombinedMultiDict([ImmutableMultiDict([])])
```
I would expect `values` to show `form` contents, even if it is an HTTP `GET`.
This used to work with Flask 1.

**Comment by ThiefMaster** [comment]
GET requests do not have a body.

**Comment by ThiefMaster** [comment]
details in this issue: https://github.com/pallets/werkzeug/pull/2037

@davidism might be useful to mention this in the flask changelog as well?

**Comment by zx80** [comment]
> GET requests should not have a body.

Sure...

 - Who cares that the standard says that, if it does not harm to do more?
 - Working transparently on args/form does helps: having a REST API client with GET means that one gets limited on the parameter sizes if they are in the URL only.
 - It is a regression, ISTM it contradicts the documentation which says that values is args+form, and it breaks my code heavily for no good reason.

**Comment by ThiefMaster** [comment]
If you need it, you can easily subclass Request and override the `values` property.

Also, see https://stackoverflow.com/a/983458/298479 on details why GET bodies are a bad idea.

FWIW, we made this change because someone reported this as a potential security problem with caching proxies.

**Comment by zx80** [comment]
> If you need it, you can easily subclass Request and override the `values` property.

I wish to avoid having to work around Flask regressions:-)

> Also, see https://stackoverflow.com/a/983458/298479 on details why GET bodies are a bad idea.

I do not buy the explanations there. From a REST API perspective based on HTTP, having to check both client and server side whether the method is GET or POST is a pain, and URL sizes are much more limited than body sizes. Morovever, the 2014 version of the HTTP protocol seems less clear about, it seems to tell that some server might chose to ignore de GET body contents, but it does not mean that Flask 2 has to do that!

> FWIW, we made this change because someone reported this as a potential security problem with caching proxies.

And create an undocumented regression in passing. I do not see the security issue. I guess that some proxy could decide to drop the body because it seems allowed by the spec, but I doubt any would really do so. Also, with a REST API I'm not sure I'd use a caching proxy, at least on the client side, it should be encrypted anyway. On the server side, I do not see why a proxy would interfere with requests.

My wish is that you do not break Flask in undocumented way, and if you do it there is a real added value, which I cannot see here.

**Comment by pgjones** [comment]
Hi @zx80,

This change is documented in the [Werkzeug major release 2.0 changelog](https://github.com/pallets/werkzeug/blob/main/CHANGES.rst#version-200). Note you can still use `request.form` for a GET request or subclass as suggested (Flask doesn't ignore the body).

As a rule I think Flask/Werkzeug should adopt secure defaults, which this does. There is therefore no desire on my part to revisit this decision.

**Comment by zx80** [comment]
Hi Phil (@pgjones)

> This change is documented in the [Werkzeug major release 2.0 changelog](https://github.com/pallets/werkzeug/blob/main/CHANGES.rst#version-200).

The regression I reported is about `value` behavior in `Flask`, and this is *not* documented AFAICS, because I looked for it. The fact that under the hood the change is related to some other class that I did not even import is pretty irrelevant to me as a user (who wrote a Flask extension).

> Note you can still use `request.form` for a GET request or subclass as suggested (Flask doesn't ignore the body).

Sure, I just have to reimplement the feature you broke.
 
> As a rule I think Flask/Werkzeug should adopt secure defaults, which this does.

I do agree on principle, obviously. However, I do not see any security implication related to actively hiding `form` from `value` on `GET`. `value` is basically a user convenience that you decided to make inconvenient so that it lost its value. The user can lose their time to work around it in 2 lines of code which reimplement what you removed in 6. What's the point, really?

> There is therefore no desire on my part to revisit this decision.

I do not understand the *therefore*, as I cannot see any improved security related to this trivial and inconvenient change, that people will just have to work around by writing more code in their app. Who cares?
